packages/core/epoch/types.go

```diff
@@ -179,3 +176,129 @@ func (e *ECRecord) FromBytes(bytes []byte) (err error) {
+	}
+
+	return a
+}
+
+// Add adds a node to the activity log.
+func (a *ActivityLog) Add(nodeID identity.ID) (added bool) {
+	return a.SetEpochs.Add(nodeID)
+
+}
+
+// Remove removes a node activity from the log.
+func (a *ActivityLog) Remove(nodeID identity.ID) (removed bool) {
+	return a.SetEpochs.Delete(nodeID)
+}
+
+// Active returns true if the provided node was active.
+func (a *ActivityLog) Active(nodeID identity.ID) (active bool) {
+	if a.SetEpochs.Has(nodeID) {
+		return true
+	}
+
+	return
+}
+
+// String returns a human-readable version of ActivityLog.
+func (a *ActivityLog) String() string {
+	var builder strings.Builder
+	builder.WriteString(fmt.Sprintf("ActivityLog(len=%d, elements=", a.SetEpochs.Size()))
+	a.SetEpochs.ForEach(func(nodeID identity.ID) {
+		builder.WriteString(fmt.Sprintf("%s, ", nodeID.String()))
+
+	})
+	builder.WriteString(")")
+	return builder.String()
+}
+
+// Clone clones the ActivityLog.
+func (a *ActivityLog) Clone() *ActivityLog {
+	clone := NewActivityLog()
+
+	a.SetEpochs.ForEach(func(nodeID identity.ID) {
+		clone.SetEpochs.Add(nodeID)
+	})
+
+	return clone
+}
+
+// Encode ActivityLog a serialized byte slice of the object.
+func (a *ActivityLog) Encode() ([]byte, error) {
+	objBytes, err := serix.DefaultAPI.Encode(context.Background(), a.SetEpochs, serix.WithValidation())
+	if err != nil {
+		// TODO: what do?
+		panic(err)
+	}
+	return objBytes, nil
+}
+
+// Decode deserializes bytes into a valid object.
+func (a *ActivityLog) Decode(data []byte) (bytesRead int, err error) {
+
+	a.SetEpochs = set.New[identity.ID]()
+	bytesRead, err = serix.DefaultAPI.Decode(context.Background(), data, &a.SetEpochs, serix.WithValidation())
+	if err != nil {
+		err = errors.Errorf("failed to parse ActivityLog: %w", err)
+		return
+	}
+
+	return
+}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// SnapshotEpochActivity is the data structure to store node activity for the snapshot.
+type SnapshotEpochActivity map[Index]*SnapshotNodeActivity
+
+// NewSnapshotEpochActivity creates a new SnapshotEpochActivity instance.
+func NewSnapshotEpochActivity() map[Index]*SnapshotNodeActivity {
+	return make(SnapshotEpochActivity)
+}
+
+// SnapshotNodeActivity is structure to store nodes activity for an epoch.
+type SnapshotNodeActivity struct {
+	model.Immutable[SnapshotNodeActivity, *SnapshotNodeActivity, nodeActivityModel] `serix:"0"`
+}
+
+// NewSnapshotNodeActivity creates a new SnapshotNodeActivity instance.
+func NewSnapshotNodeActivity() *SnapshotNodeActivity {
+	return model.NewImmutable[SnapshotNodeActivity](&nodeActivityModel{NodesLog: make(map[identity.ID]uint64)})
+}
+
+// nodeActivityModel stores node identities and corresponding accepted block counters indicating how many blocks node issued in a given epoch.
+type nodeActivityModel struct {
+	NodesLog map[identity.ID]uint64 `serix:"0,lengthPrefixType=uint32"`
```

Comment:
in case blockOrphaned event will happen we will know if it was the only block issued in this epoch, so we will know node should be removed from activity log